src/service_mongoose_system_stats.erl

```diff
@@ -0,0 +1,209 @@
+    ok;
+handle_event(_StatType, _Map1, _Map2, _Config) ->
+    ok.
+
+parse_telemetry_report(ClientId, #{module := Module}, #{host := _Host, opts := Opts}) ->
+    Backend = proplists:get_value(backend, Opts, none),
+    build_report(ClientId, modules, Module, Backend);
+parse_telemetry_report(ClientId, #{hosts_count := HostsCount}, _Metadata) ->
+    build_report(ClientId, hosts_count, HostsCount);
+parse_telemetry_report(ClientId, #{event_category := EventCategory}, #{event_action := EventAction}) ->
+    build_report(ClientId, EventCategory, EventAction);
+parse_telemetry_report(ClientId, #{event_category := EventCategory}, #{event_action := EventAction, event_label := EventLabel}) ->
+    build_report(ClientId, EventCategory, EventAction, EventLabel);
+parse_telemetry_report(_ClientId, _Metrics, _Metadata) ->
+    %incorrect reports are ignored
+    ok.
+
+handle_info(flush_reports, State = #state{reports = Reports, loop_timer_ref = TimerRef, report_after = ReportAfter}) ->
+    UrlBase = maybe_get_url(),
+    erlang:cancel_timer(TimerRef),
+    flush_reports(UrlBase, Reports),
+    NewTimerRef = erlang:send_after(ReportAfter, self(), flush_reports),
+    {noreply, State#state{reports = [], loop_timer_ref = NewTimerRef}};
+handle_info(_Message, _State) ->
+    ok.
+
+maybe_get_url() ->
+    MaybeUrl = ejabberd_config:get_local_option(google_analytics_url),
+    get_url(MaybeUrl).
+get_url(undefined)->
+    ?BASE_URL;
+get_url(Url) ->
+    Url.
+
+handle_cast({add_report, {Metrics, Metadata}}, State = #state{reports = Reports, client_id = ClientId}) ->
+    NewReport = parse_telemetry_report(ClientId, Metrics, Metadata),
+    FullReport = [NewReport | Reports],
+    maybe_flush_report(length(FullReport)),
+    {noreply, State#state{reports = FullReport}}.
+
+% %%-----------------------------------------
+% %% Helpers
+% %%-----------------------------------------
+
+report_hosts_count() ->
+    Hosts = ejabberd_config:get_global_option(hosts),
+    NumberOfHosts = length(Hosts),
+    telemetry:execute(?STAT_TYPE, #{hosts_count => NumberOfHosts}, #{}).
+
+maybe_flush_report(ReportLength) when ReportLength >= 20 ->
+    self() ! flush_reports;
+maybe_flush_report(_) ->
+    ok.
+
+% % https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide#batch-limitations
+% % A maximum of 20 hits can be specified per request.
+flush_reports(_, []) ->
+    ok;
+flush_reports(ReportUrl, Lines) when length(Lines) =< 20 ->
+    Headers = [],
+    ContentType = "",
+    Body = string:join(Lines, "\n"),
+    Request = {ReportUrl, Headers, ContentType, Body},
+    httpc:request(post, Request, [], []);
+flush_reports(ReportUrl, Lines) ->
+    {NewBatch, RemainingLines} = lists:split(20, Lines),
+    flush_reports(ReportUrl, NewBatch),
+    flush_reports(ReportUrl, RemainingLines),
+    ok.
+
+get_client_id() ->
+    get_client_id(20).
+get_client_id(0) ->
+    no_client_id;
+get_client_id(Counter) when Counter > 0 ->
```

Comment:
Now, this function is clean and neat, but maybe we can save the redundancy of checking the table twice, there's still a bit of duplicated work between this one and `maybe_make_and_save_new_client_id/0`. Any ideas? It's all about pretty code here 🙂

Reply:
I wouldn't worry about this part of the code being pretty as it is to be modified in the next iterations. Client ID will be saved to the file and read once when the service is starting up.